### PR TITLE
Codecov & Botonic workflows

### PR DESCRIPTION
**Pull request by manuelfidalgo**
## Description

1. **Setup codecov** for botonic:
    - Configure initial codecov setup
    - Upload coverage report to https://codecov.io/gh/hubtype/botonic
2. Trigger **Github workflows only when necessary**. 

## Context
1. Codecov
Configuring codecov will allow us to improve the coverage on every push (or at least keep it as the level of previous push)
The configuration in [codecov.yml](https://docs.codecov.io/docs/codecovyml-reference) includes an initial setup with the following:
    - Minimum coverage of 40% threshold
    - [Flags](https://docs.codecov.io/docs/flags) for the different projects in packages/ that have unit tests running. (Flags help us isolating reports in a monorepo setup)
You can see the reports in https://codecov.io/gh/hubtype/botonic

2. Trigger Github workflows only when necessary 
Prior to this PR, all workflows were triggered, no matter where the change was coming from.
Now, each workflow is triggered only when the change is originated in the path where the workflow operates. For instance, changes in packages/botonic-plugin-dynamodb/** will trigger botonic-plugin-dynamo-tests.yml workflow only. Of course, if we change a workflow file, it will be executed once it is pushed.

## Approach taken / Explain the design
1. Codecov
    - Include a codecov config file with the initial setup
    - add codecov/codecov-action@v1 to worflows in order to upload coverage reports
2. Trigger Github workflows only when necessary 
    - Include paths section on the trigger pointing to the related codebase in the packages folder. Also it makes reference to the workflow itself, to make sure it is executed if changes on it are pushed.

## To document / Usage example
You can see last generated reports at https://codecov.io/gh/hubtype/botonic

## Testing
I executed some tests to see that results from the different paths appear in the codecov page.
Regarding the event that triggers the workflows, I also did some tests but, as I have changed all workflows and paths sometimes are hard to validate, something might have fallen through the cracks, please let me know if you see that workflows are not being triggered upon your changes.

### CODE REVIEW

**Review by ericmarcos — APPROVED** [review]
Nice!! 👏👏👏

File: .github/workflows/botonic-cli-tests.yml [hunk 1]
```diff
@@ -1,6 +1,10 @@
 name: Botonic cli tests
 
-on: [push]
+on:
+  push:
+    paths:
+      - 'packages/botonic-cli/**'
+      - '.github/workflows/botonic-cli-tests.yml'
```
Comment (dpinol) [inline .github/workflows/botonic-cli-tests.yml:7]: are you sure that you need to include the workflow file itself?

File: .github/workflows/botonic-plugin-contentful-tests.yml [hunk 2]
```diff
@@ -3,11 +3,8 @@ name: Botonic plugin-contentful tests
 on:
   push:
     paths:
-      - '*'
-      - '.github/**'
-      - 'packages/*'
       - 'packages/botonic-plugin-contentful/**'
```
Comment (dpinol) [inline .github/workflows/botonic-plugin-contentful-tests.yml:6]: please restore '*'' and 'packages/*' There are there build configurations

File: .github/workflows/botonic-plugin-dashbot-tests.yml [hunk 3]
```diff
@@ -21,8 +27,8 @@ jobs:
         restore-keys: |
           ${{ runner.os }}-node-
     - name: Install botonic-plugin-dashbot (Step 4 of 6)
-      run: (cd ./packages/botonic-plugin-dashbot && npm install -D)
+      run: (cd ./packages/$PACKAGE && npm install -D)
```
Comment (dpinol) [inline .github/workflows/botonic-plugin-dashbot-tests.yml:30]: :clap:

File: .github/workflows/botonic-plugin-dashbot-tests.yml [hunk 4]
```diff
@@ -21,8 +27,8 @@ jobs:
...
     - name: Install common packages dependencies (Step 5 of 6)
       run: npm install -D
     - name: Verify lint botonic-plugin-dashbot (Step 6 of 6)
```
Comment (dpinol) [inline .github/workflows/botonic-plugin-dashbot-tests.yml:33]: please remove "botonic-plugin-dashbot" from the name ( I already removed it from most fpackages, but please check if there are more), so that the workflow files are most similar

File: .github/workflows/botonic-react-tests.yml [hunk 5]
```diff
@@ -31,4 +35,12 @@ jobs:
     - name: Verify lint
       run: (cd ./packages/$PACKAGE && npm run lint_ci)
     - name: Verify index.d.ts
-      run: scripts/qa/lint-d-ts.sh packages/botonic-core
+      run: scripts/qa/lint-d-ts.sh ./packages/$PACKAGE
+    
+    - name: Upload coverage to codecov
```
Comment (dpinol) [inline .github/workflows/botonic-react-tests.yml:40]: can you try this example https://stackoverflow.com/a/59798779/145289  to see if you can have the codecov config in a single file?

**Review by dpinol — CHANGES_REQUESTED** [review]
:clap:

File: .github/workflows/botonic-cli-tests.yml (hunk 1 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-cli-tests.yml:7]: From what I understood from the [docs](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths), a git diff is performed so I think it is needed.
I think I did a test on that but I will double check

File: .github/workflows/botonic-plugin-contentful-tests.yml (hunk 2 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-plugin-contentful-tests.yml:8]: ok, sorry, I thought they were not needed. I pushed a new commit with the changes

**Review by vanbasten17 — APPROVED** [review]
Nice!! 💯 go for it after taking a look at @dpinol suggestions!

File: .github/workflows/botonic-react-tests.yml (hunk 5 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-react-tests.yml:40]: From what I have read (https://github.com/actions/runner/issues/646), that would the case for reusing a pure bash/docker/JS action, not to call one action from another.
I didn't tried because of that link but I can give it a try.

File: .github/workflows/botonic-plugin-dashbot-tests.yml (hunk 4 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-plugin-dashbot-tests.yml:33]: I created https://github.com/hubtype/botonic/pull/1177/commits/ba06acede628229d3aee2fe5e04eff27f8bcba85 for this

File: .github/workflows/botonic-plugin-contentful-tests.yml [hunk 6]
```diff
@@ -4,10 +4,9 @@ on:
   push:
     paths:
       - '*'
```
Comment (dpinol) [inline .github/workflows/botonic-plugin-contentful-tests.yml:6]: sorry, can you add '*' and 'packages/*' for all packages?

File: .github/workflows/botonic-nlu-tests.yml [hunk 7]
```diff
@@ -1,11 +1,17 @@
...
  
 jobs:
   botonic-plugin-nlu-tests:
     name: Botonic nlu tests
     runs-on: ubuntu-latest
+    env:
+      PACKAGE: botonic-nlu
     steps:
     - name: Checking out to current branch (Step 1 of 7)
       uses: actions/checkout@v1
```
Comment (dpinol) [inline .github/workflows/botonic-nlu-tests.yml:17]: make the most of this PR to try to upgrade to actions/checkout@v2, actions/cache@v2 and actions/setup-node@v2
luckily you can do a replace-all for this

**Review by dpinol — CHANGES_REQUESTED** [review]

File: .github/workflows/botonic-cli-tests.yml (hunk 1 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-cli-tests.yml:7]: I did some checks on https://github.com/hubtype/botonic/tree/chore/codecov-temp branch and I confirm that line is needed: if we remove it, changes on the .yml file won't trigger the workflow.

File: .github/workflows/botonic-nlu-tests.yml (hunk 7 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-nlu-tests.yml]: updated in https://github.com/hubtype/botonic/pull/1177/commits/6076e6a62da605e85b0a648245c0014480f2ac7e

File: .github/workflows/botonic-plugin-contentful-tests.yml (hunk 6 above)
Comment (manuelfidalgo) [reply .github/workflows/botonic-plugin-contentful-tests.yml:6]: included in https://github.com/hubtype/botonic/pull/1177/commits/7e317273c5bb55d5d76562a61db9d4b21bf14d58

**Review by dpinol — APPROVED** [review]
:clap:

File: .github/workflows/botonic-cli-tests.yml [hunk 8]
```diff
@@ -1,6 +1,12 @@
 name: Botonic cli tests
 
-on: [push]
+on:
+  push:
+    paths:
+      - '*'
```
Comment (gbarba) [inline .github/workflows/botonic-cli-tests.yml:6]: I guess that a single `*` is not recursive so it doesn't means a change anywhere

File: .github/workflows/botonic-cli-tests.yml [hunk 9]
```diff
@@ -1,6 +1,12 @@
 name: Botonic cli tests
 
-on: [push]
+on:
```
Comment (gbarba) [inline .github/workflows/botonic-cli-tests.yml:3]: You are not uploading the coverage for this package.
I see it isn't running any test (there isn't any `npm run test`).
I would prefer to also upload this coverage that will be 0%, but this will make visible that we don't have any testing for this package.
But this can go in a separate PR to get this merged

**Review by gbarba — APPROVED** [review]
I would like to see another PR that adds the codecov upload for all packages so we make explicit that we have these packages untested (and we can see progress in the future).
This, I guess, will require to add a `npm run test` to them, despite this will run nothing.
But confirm with @dpinol and botonic team

Also I see lot of repetition. maybe we can think in a refactor to put a "generic workflow" that is called with params for each workflow
